fix(platform-browser): key event handler should run in the correct zone

JiaLiPassion opened the pull request:
Close #38260 .

Currently the KeyEventPlugin handle key event such as `keydown.enter` and always run into `ngZone` no matter
which zone it is added. And this behavior is not consistent with the other event handlers.

So this commit fixes this issue.

Also it make sure the `event handler` in key event plugin call on the correct context object.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: #38260

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

mhevery commented on packages/platform-browser/test/dom/events/key_events_spec.ts:
```diff
@@ -66,5 +71,80 @@ import {KeyEventsPlugin} from '@angular/platform-browser/src/dom/events/key_even
+      const doc = getDOM().supportsDOMEvents() ? document : getDOM().createHtmlDocument();
+      const manager = new EventManager([new KeyEventsPlugin(doc)], zone);
+
+      it('key event handler should in the ngZone when added inside ngZone', () => {
```
```suggestion
      it('key event handler should be in the ngZone when added inside ngZone', () => {
```

mhevery commented on packages/platform-browser/test/dom/events/key_events_spec.ts:
```diff
@@ -66,5 +71,80 @@ import {KeyEventsPlugin} from '@angular/platform-browser/src/dom/events/key_even
+        expect(isInNgZone).toEqual(`key handler inside ngZone: true`);
+      });
+
+      it('global key event handler should in the ngZone when added inside ngZone', () => {
```
```suggestion
      it('global key event handler should be in the ngZone when added inside ngZone', () => {
```

mhevery commented on packages/platform-browser/test/dom/events/key_events_spec.ts:
```diff
@@ -66,5 +71,80 @@ import {KeyEventsPlugin} from '@angular/platform-browser/src/dom/events/key_even
+        expect(isInNgZone).toEqual(`key handler inside ngZone: true`);
+      });
+
+      it('key event handler should outside of the ngZone when added outside of the ngZone', () => {
```
```suggestion
      it('key event handler should be outside of the ngZone when added outside of the ngZone', () => {
```

mhevery commented on packages/platform-browser/test/dom/events/key_events_spec.ts:
```diff
@@ -66,5 +71,80 @@ import {KeyEventsPlugin} from '@angular/platform-browser/src/dom/events/key_even
+        expect(isInNgZone).toEqual(`key handler inside ngZone: false`);
+      });
+
+      it('global key event handler should outside of the ngZone when added outside of the ngZone',
```
```suggestion
      it('global key event handler should be outside of the ngZone when added outside of the ngZone',
```

mhevery commented on packages/platform-browser/testing/src/browser_util.ts:
```diff
@@ -116,6 +375,58 @@ export function createMouseEvent(eventType: string): MouseEvent {
+  if (!key) {
+    return -1;
+  }
+  for (let i = 0; i < keyCodeMapping.length; i++) {
```
instead of having a big loop, would it not be more intuitive to store `keyCodeMapping` as a map?

mhevery commented on packages/platform-browser/testing/src/browser_util.ts:
```diff
@@ -116,6 +375,58 @@ export function createMouseEvent(eventType: string): MouseEvent {
+  return -1;
+}
+
+export function createKeyboardEvent(
```
This seems like a public API. Can we have proper jsdocs for it?

mhevery commented on packages/platform-browser/src/dom/events/key_events.ts:
```diff
@@ -167,13 +164,12 @@ export class KeyEventsPlugin extends EventManagerPlugin {
-  static eventCallback(fullKey: any, handler: Function, zone: NgZone): Function {
-    return (event: any /** TODO #9100 */) => {
+  static eventCallback(fullKey: any, handler: Function): Function {
+    return function(this: unknown, event: any /** TODO #9100 */) {
```
could we resolve the todo?

jessicajaniuk commented:
@JiaLiPassion Any updates on this PR? Is it still needed?

devversion commented on packages/platform-browser/src/dom/events/event_manager.ts:
```diff
@@ -47,7 +47,16 @@ export class EventManager {
   addEventListener(element: HTMLElement, eventName: string, handler: Function): Function {
     const plugin = this._findPluginFor(eventName);
-    return plugin.addEventListener(element, eventName, handler);
+    if (NgZone.isInAngularZone()) {
```
How helpful is it actually knowing that the event is "registered" in the proper zone? Couldn't the event plugins still somehow mess this up, depending on how the event is registered (it might not be patched by zonejs for e.g. hammerjs right?)

TL;DR: Would it make sense to always run the handler guarded in the zone?

JiaLiPassion replied:
@devversion, I updated the PR to include several fixes.
And the reason I didn't use 
```
ngZone.runOutside(() => {
  addEventListener(element, eventName, () => {
     ngZone.run(() => {listener(eventName});
  });
});
```
to do this because there are several issues with this solution.

1. this solution will bypass all functionality of `zone.js`, zone.js will not be able to moniter the `eventTask`, so the `onScheduleTask`, `onInvokeTask` will not be triggered if some user want to moniter.
2. now `zone.js` support `UNPATCHED_EVENTS` feature, the user can define `UNPATCHED_EVENTS=['mouseover', 'scroll']` so the `zone.js` will use the `native addEventListener` for these events, and the listeners will not trigger change detection. So if we are using the solution above, this feature will be lost.

So to still use `zone.js` functionality, the behavior should be ensure the `addEventListener` happened inside of `NgZone`.

devversion replied:
@JiaLiPassion all of this makes sense! I think my comment was a little ambiguous/unclear. I was more asking why we would not always do the following:

```js
ngZone.run(() => 
  plugin.addEventListener(element, eventName, handlerEnsureInZone));
```

I added a new comment for it though. Thanks for the reply/and details.

devversion reviewed (dismissed):
LGTM conceptually. Nice 👍 

@JiaLiPassion should we also remove the `runGuarded` calls in other event plugins then?

dylhunn reviewed (approved):
reviewed-for: fw-core

JiaLiPassion committed:
fix(platform-browser): listener in the template should always in NgZone

When the user declare `event listener` in the angular template, the user
expect the `listner` run inside `NgZone` and trigger the change
detection automatically, but if the `addEventListener` is running
outside of `NgZone`, the `listener` will run outside of `NgZone`.
For example, the user may render the component dynamically outside
of `NgZone` which will be this case.

So this commit ensure the `event listener` is added inside of `NgZone`.

JiaLiPassion committed:
feat(platform-browser): bind __ngUnwrap__ as function property

Now in the `DebugNode.triggerEventListener()`, we are using
`function.toString()` contains `__ngUnwrap__` to check whether
the event listener can get the unwrapped listener, which is not
an elegent solution, since if the user wrap the `listener` again
during `addEventListener` (such as `KeyEventManagerPlugin`),
`listener.toString()` will not contain `__ngUnwrap__` any longer.

So this commit attach the unwrapped listener as the wrapped listener's
`__ngUnwrap__` property.

JiaLiPassion committed:
fix(core): should always run ApplicationRef.tick() inside the NgZone

`ApplicationRef` monitors the `NgZone.onStable` and run `tick()` inside
`NgZone`, but if the apps calls the `ApplicationRef.tick()` themselves
outside of the `ngZone`, it may trigger the re-entry of the `tick()`.
The process is:

1. `Application.tick()` run outside of the `NgZone`.
2. `tick()` run change detection and register the event listener.
3. the event listener will be registered inside `NgZone` by
`NgZone.run()`.
4. The `NgZone.run()` triggers another `onStable` emit and triggers
the `ApplicationRef.tick()` again.

So this commit ensure `ApplicationRef.tick()` always inside the
`NgZone`.

devversion commented on packages/core/src/application_ref.ts:
```diff
@@ -831,6 +829,15 @@ export class ApplicationRef {
+    if (NgZone.isInAngularZone()) {
+      this._tick();
+    } else {
+      this._zone.run(() => this._tick());
```
This is related to my previous comment (which I think was a little confusingly phrased): Is there any benefit in checking with `NgZone.isInAngularZone()`, or could we just always run it in the zone? like:

```ts
function tick(): void {
  this._zone.run(() => this._tick());
}
```

JiaLiPassion replied:
@devversion , the reason not to use `this._zone.run()` directly is if we use `this.zone.run()` even we are already inside the `NgZone`, it will trigger another `ApplicationRef.tick()` and cause the `NG0101 nested ApplicationRef.tick()` error.

devversion replied:
interesting, good thing to have this noted here 👍

devversion commented on packages/platform-browser/src/dom/dom_renderer.ts:
```diff
@@ -55,7 +55,7 @@ function decoratePreventDefault(eventHandler: Function): Function {
+  const listener: Function = (event: any) => {
     // Ivy uses '__ngUnwrap__' as a special token that allows us to unwrap the function
     // so that it can be invoked programmatically by `DebugNode.triggerEventHandler`. The debug_node
     // can inspect the listener toString contents for the existence of this special token. Because
```
Can this logic below `event === '__ngUnwrap__'` be removed now?

JiaLiPassion replied:
I am just concerned about that some app or the dev tools are using `eventListener('__ngUnwrap__')` to retrive the original listener, I will remove it and if the `g3` is ok, I think it is safe.

devversion replied:
👍

devversion commented on packages/platform-browser/src/dom/events/event_manager.ts:
```diff
@@ -108,4 +108,15 @@ export abstract class EventManagerPlugin {
     return this.addEventListener(target, eventName, handler);
   }
+
+  ensureAddListenerInZone(element: HTMLElement, eventName: string, handler: Function): Function {
```
If you would be okay with it: I think we should add a `JSDoc` description here explaining why this is desirable. It's surely obvious to us from the context of your PR but this would be helpful long-term I feel like

JiaLiPassion replied:
Sure, I just add the `JSDoc` to explain this function.

devversion reviewed (commented):
Seems very reasonable to me. Just added one question (similar to my previous) comment, wondering if we could simplify code a little more

dylhunn reviewed (dismissed):
reviewed-for: fw-core

AndrewKushnir commented:
[Global Presubmit](https://test.corp.google.com/OCL:431828588:BASE:431828730:1646190511323:68bf6ddf).

AndrewKushnir commented:
FYI, based on a global presubmit, there are ~500 failing targets, so I'm adding the "blocked" label for now.

JiaLiPassion committed:
test
